core/src/main/scala/kafka/coordinator/group/GroupCoordinator.scala

```diff
@@ -338,6 +338,28 @@ class GroupCoordinator(val brokerId: Int,
     }
   }
 
+  def handleDeleteGroups(groupIds: Set[String]): Map[String, Errors] = {
+    if (!isActive.get) {
+      groupIds.map((_ -> Errors.COORDINATOR_NOT_AVAILABLE)).toMap
+    } else {
+      var groupErrors: Map[String, Errors] = Map()
+      var eligibleGroups: Seq[String] = Seq()
+
+      groupIds.foreach { groupId =>
```

Comment:
In fact, this is also a "check and act." It is possible for an eligible group to be unloaded between these checks and the call to `deleteGroups`. 

I think we should follow a structure more similar to the other API handlers. I would suggest moving the state checking that we currently have in `GroupMetadataManager.deleteGroups` into the `else` case below. We should do the following:

1. Check if there is no group or if the group is Dead. If so, it could mean that it has already been moved to another broker or it could mean that the group doesn't exist. I am not sure we have a bulletproof way to distinguish these cases, but maybe we could just check again if the coordinator is still correct? 
2. Check if the group is not empty. If so, return the GROUP_NOT_EMPTY error code.
3. If the group is empty, we should transition to `Dead`. Once we do so, we are ensured that no other thread will attempt to use the `GroupMetadata` object. We can then collect this eligible group in a collection (as is currently done) and send it to `cleanupGroupMetadata` outside of the lock.

Reply:
I'm working on this and have a couple of questions for now:
1. It seems all this can be done here and we could get rid of `GroupMetadataManager.deleteGroups()`. Do you see an issue with it?
2. Could you please clarify what you mean by "check again if the coordinator is still correct" when group cannot be found or is `Dead`?

Thanks.

Reply:
1. Seems reasonable to me.

2. We are trying to address the case in which a group gets deleted or migrated in between the time that we check if the coordinator is assigned and we delete the group metadata. We have the `Dead` state for this purpose, so whenever we check `GroupMetadata`, the first thing we should check is whether it is already `Dead`. If it is, then we know it was either already deleted or already migrated. My suggestion is to check again whether we are still the coordinator for the group to disambiguate the two cases. 

Note that I am not sure that this is 100% bulletproof. For example, it may not handle the case when the coordinator is migrated away and then back very quickly. A spurious NOT_COORDINATOR error is not a big deal because clients are expected to handle it, but I am not too sure about the GROUP_ID_NOT_FOUND error. Maybe clients just have to treat it with the same skepticism that they treat the UNKOWN_TOPIC_OR_PARTITION errors.

Reply:
Thanks for clarifying this. I'll try to do just that in the new patch. Will submit shortly.